src/Storages/StorageS3.h

```diff
@@ -215,8 +266,22 @@ class StorageS3Source : public SourceWithKeyCondition, WithContext
         explicit operator bool() const { return reader != nullptr; }
         PullingPipelineExecutor * operator->() { return reader.get(); }
         const PullingPipelineExecutor * operator->() const { return reader.get(); }
-        String getPath() const { return fs::path(bucket) / key_with_info->key; }
-        const String & getFile() const { return key_with_info->key; }
+        String getPath() const
+        {
+            return key_with_info->path_in_archive.has_value()
+                ? (bucket + "/" + key_with_info->key + "::" + key_with_info->path_in_archive.value())
+                : (bucket + "/" + key_with_info->key);
```

Comment:
```suggestion
            return bucket + "/" + key_with_info->getPath();
```
This should be identical